docs/_Reproducible-Data-Analysis/03-software-containers.md

```diff
@@ -5,3 +5,86 @@ layout: default
+
+- **Use Specific Versions**: Specify exact versions of base images, software, and libraries, in order to avoid breaking changes occuring when updating with the `latest` tag and ensures consistency across environments.
+
+- **Minimize Layer Size**: Structure your definition file to combine related commands into single layers to reduce the container size which speeds up download and deployment.
+
+- **Clean Up**: Remove unnecessary packages and clear cache in the same layer where installations occur to minimize the container's footprint.
+
+- **Non-root User**: Run the container as a non-root user whenever possible, which enhances the security of the container, reducing the risk of privilege escalation attacks.
+
+- **Base Image Selection**: Choose a minimal base image that includes only the necessary packages and libraries for your application, to minimizes the attack surface and the container size.
+
+- **Immutable Containers**: Treat containers as immutable.
+For updates or changes, build a new container image.
+This facilitates modularity and version control while ensuring reproducibility.
+
+- **Security Scanning**: Regularly scan your containers for vulnerabilities and apply patches as needed.
+Keeping your containers updated is crucial for security.
+
+- **Efficient Data Management**: Store data and logs outside of containers to ensure persistence and scalability.
+Use volumes or bind mounts for data that needs to persist beyond the life of the container.
+
+- **Documentation**: Include a `%help` section in your definition file, providing users with information on how to use the container, including running the software and accessing data.
+
+
+## Advanced Usage
+#### [Integration with Nextflow](https://www.nextflow.io/docs/latest/container.html)
+- **Nextflow and Containers**: Simplifies complex workflows by executing each step in a container for consistency across environments.
+- **Configurations**: Supports managing containers through `nextflow.config`, streamlining execution.
+
+#### [Kubernetes and Containers](https://kubernetes.io/docs/home/)
+- **Container Orchestration**: Automates deployment, scaling, and management of containerized applications, essential for microservices architecture.
+- **Scalability and Management**: Provides tools for load balancing, auto-scaling, and efficient resource allocation across diverse infrastructures.
+
+## Resources and Further Reading
+- [Apptainer User Guide](https://apptainer.org/docs/user/latest/introduction.html): Comprehensive documentation for getting started with Apptainer.
+- [BioContainers Community](https://biocontainers.pro/): A resource for finding and sharing containerized bioinformatics tools.
+- [Docker Introduction @Carpentries](https://carpentries-incubator.github.io/docker-introduction/)
```

Comment:
maybe mention that the lesson  is not endorsed by The Carpentries as it is still in the incubator

```suggestion
- [Docker Introduction Lesson (Beta version)](https://carpentries-incubator.github.io/docker-introduction/)
```